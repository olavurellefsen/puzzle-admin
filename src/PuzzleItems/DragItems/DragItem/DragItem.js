import React, { useState } from "react";
import {
  DragItemContainer,
  DragItemForm,
  DragItemTopContent,
  DragItemImage,
  DragItemBubbleText,
  DragItemFieldCaption,
  DragItemField,
  DragItemRestContent
} from "./DragItem.style";
import ToggleSwitch from "../../../Fields/ToggleSwitch/ToggleSwitch";
import gql from "graphql-tag";
import { Mutation } from "react-apollo";

export default props => {
  const { index, dragitem } = props;
  const wait = dragitem.wait;
  const [ waitField, setWaitField ] = useState(wait);
  const puzzleItem = dragitem.puzzleItemBypuzzleItemId;
  const item = puzzleItem.itemBybubbleTextId;

  // If the parent component is updated (someone else has changed the value), then update the field here
  if (wait !== waitField) {
    setWaitField(wait);
  }

  const UPDATE_DRAGITEM = gql`
    mutation update_dragitem($id: Int!, $wait: Boolean!) {
      update_dragitem(where: { id: { _eq: $id } }, _set: { wait: $wait }) {
        affected_rows
      }
    }
  `;
  
  return (
    <DragItemContainer
      data-testid="DragItemContainer"
      className={`drag${dragitem.id}`}
      id={`drag${dragitem.id}`}
    >
      <Mutation mutation={UPDATE_DRAGITEM}>
        {(updateDragItem, { data }) => (
          <DragItemForm
            onSubmit={e => {
              e.preventDefault();
              updateDragItem({
                variables: { id: dragitem.id, wait: waitField }
              })
            }}>
            <DragItemTopContent>
              <DragItemImage
                key={index}
                src={`images/puzzleitems/${puzzleItem.imagefile}`}
                data-testid="DragItemsImage"
              />
              <DragItemBubbleText>
                <DragItemFieldCaption>Bubble text</DragItemFieldCaption>
                <DragItemField>{item.value}</DragItemField>
              </DragItemBubbleText>
            </DragItemTopContent>
            <DragItemRestContent>
              <DragItemFieldCaption>Name</DragItemFieldCaption>
              <DragItemField>{puzzleItem.name}</DragItemField>
              <DragItemFieldCaption>Image file</DragItemFieldCaption>
              <DragItemField>{puzzleItem.imagefile}</DragItemField>
              <DragItemFieldCaption>Wait</DragItemFieldCaption>
              <ToggleSwitch
                toggled={waitField}
                onToggle={() => {
                  updateDragItem({
                    variables: { id: dragitem.id, wait: !waitField }
                  })
                  setWaitField(!waitField);
                }}
              />
            </DragItemRestContent>
          </DragItemForm>
        )}
      </Mutation>
    </DragItemContainer>
  );
};
