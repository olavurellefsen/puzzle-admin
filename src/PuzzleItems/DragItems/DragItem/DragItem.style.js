import styled from "styled-components";

export const DragItemContainer = styled.div`
  height: 180px;
  width: 240px;
  padding: 10px;
  margin: 20px;
  display: flex;
  flex-direction: column;
  box-shadow: 0 4px 8px 0 rgba(0,0,0,${props => (props.selected ? 0.9 : 0.7)});
  transition: 0.3s;
  &:hover {
    box-shadow: 0 8px 16px 0 rgba(0,0,0,0.8);
  }
  opacity: ${props => (props.selected ? 1 : 0.8)};
  background-image: url('${props => props.backgroundImage}');
  background-size: cover;
`;

export const DragItemForm = styled.form`
  display: flex;
  flex-direction: column;
  padding: 2px;
`;

export const DragItemFieldCaption = styled.div`
  font-size: 12px;
  font-weight: normal;
  color: gray;
`;

export const DragItemInputField = styled.input`
  font-size: 14px;
  padding: 0px 0px 10px 0px;
  background-color: rgba(0, 0, 0, 0);
  border: 0px;
`;