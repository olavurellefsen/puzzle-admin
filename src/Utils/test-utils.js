import React from "react";
import { render } from "@testing-library/react";
import { ApolloProvider } from "react-apollo";
import MainContextStore from "../MainContextStore";
import client from "../heyBreydLocalizationClient";

const AllTheProviders = ({ children }) => {
  return (
    <ApolloProvider client={client}>
      <MainContextStore>{children}</MainContextStore>
    </ApolloProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
