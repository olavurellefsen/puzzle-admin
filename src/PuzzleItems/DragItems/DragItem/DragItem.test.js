import React from "react";
import { render, cleanup } from 'test-utils';
import DragItem from "./DragItem";
afterEach(cleanup)

it("should render correctly", () => {
  const { container } = render(<DragItem
    index={0}
    dragitem={{id: 0, puzzleItemBypuzzleItemId:
      { name: "dummy name",
        imagefile: "dummyimage.png",
        textitemBynameTextitemId: { id: 1, textitemLanguagesBytextitemId : [{ textvalue: "name"}]},
        textitemBypuzzletextTextitemId: { id: 2, textitemLanguagesBytextitemId : [{ textvalue: "puzzletext"}]},
      }
    }}/>);
  expect(container).toMatchSnapshot();
});

it("should show the DragItemContainer element", () => {
  const { queryByTestId } = render(<DragItem
    index={0}
    dragitem={{id: 0, puzzleItemBypuzzleItemId:
      { name: "dummy name",
        imagefile: "dummyimage.png",
        textitemBynameTextitemId: { id: 1, textitemLanguagesBytextitemId : [{ textvalue: "name"}]},
        textitemBypuzzletextTextitemId: { id: 2, textitemLanguagesBytextitemId : [{ textvalue: "puzzletext"}]},
      }
    }}/>);
  expect(queryByTestId("DragItemContainer")).toBeTruthy();
});