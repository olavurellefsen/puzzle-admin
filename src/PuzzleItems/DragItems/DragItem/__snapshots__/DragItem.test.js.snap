// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <div
    class="drag0 sc-bdVaJa fBOIxN"
    data-testid="DragItemContainer"
    id="drag0"
  >
    <form
      class="sc-bwzfXH iYKgxb"
    >
      <div
        class="sc-htpNat ibokuz"
      >
        <img
          class="sc-bxivhb cgqVai"
          data-testid="DragItemsImage"
          src="images/puzzleitems/dummyimage.png"
        />
        <div
          class="sc-ifAKCX JBEal"
        >
          <div
            class="sc-EHOje cKywGb"
          >
            Bubble text
          </div>
          <div
            class="sc-bZQynM NSNHQ"
          >
            dummy bubble text
          </div>
        </div>
      </div>
      <div
        class="sc-gzVnrw fXgyNi"
      >
        <div
          class="sc-EHOje cKywGb"
        >
          Name
        </div>
        <div
          class="sc-bZQynM NSNHQ"
        >
          dummy name
        </div>
        <div
          class="sc-EHOje cKywGb"
        >
          Image file
        </div>
        <div
          class="sc-bZQynM NSNHQ"
        >
          dummyimage.png
        </div>
        <div
          class="sc-EHOje cKywGb"
        >
          Wait
        </div>
        <div
          class="sc-htoDjs ilsOVI"
        >
          <div
            class="sc-dnqmqq hYKWMG"
            data-testid="toggle"
            width="40"
          >
            <div
              class="sc-iwsKbI hCqYwb"
              width="40"
            />
            <div
              class="sc-gZMcBi gLIrsm"
              width="40"
            />
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
